### PR TITLE
felix/bpf: pod->svc->self does not need allow policy

### PR DESCRIPTION
**Pull request by tomastigera**
We allow this traffic to match iptables dp and not require to have a special policy that allows traffic based on its host IP which is hard to express when pods start and go.

The check is done after we parsed the packet and we do it before policy and before CT. So we only do it for the first packet of a connection with minimal overhead.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
ebpf: When a pod connects via a service to self, ingress traffic is policed as if it's source is the pod and not the host after MASQ
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Review by fasaxc — COMMENTED** [review]
I think in iptables mode we do apply policy but the policy sees `srcIP=dstIP=podIP`.  That "just works" because filter.FORWARD is after DNAT but before SNAT.

Auto-allowing this traffic is better than requiring a policy (matches the CTLB behaviour) but might be best to be consistent with iptables if we can do it easily.  One fewer gotcha for E2Es and whatnot to worry about.

**Comment by tomastigera** [comment]
> I think in iptables mode we do apply policy but the policy sees `srcIP=dstIP=podIP`. That "just works" because filter.FORWARD is after DNAT but before SNAT.

That makes sense, unfortunately, in our case we do the MASQ before the ingress policy back into the pod. We do not know the original src IP by that time. We may try to redirect on the same interface using BPF_F_INGRESS and do the MASQ there ourselves. It would go through the policy then. But it is untested.

**Comment by tomastigera** [comment]
I think we can just do the policy against our own source IP as we know that the packet is from ourselves. We cannot restore the port but lets assume that it is most likely a random source port so it does not matter that we use the port after MASQ if it was changed.

We could use kfunc to lookup the MASQ conntrack and lookup the source port too, but that would only work on kernels newer than 5.18. Something we can do in a future iteration if the IP alone is a problem.

**Comment by tomastigera** [comment]
Seems like not creating policy that would allow MASQ access to self, also found an issue in how we handle host access to local pods https://github.com/projectcalico/calico/pull/6949/commits/fd732e2f5512dbbc6eaad84797a96bd9b08a1822

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Review by fasaxc — CHANGES_REQUESTED** [review]
Looks like `calico_tc_skb_accepted` accesses state->ct_result` _before_ jumping to `calico_tc_skb_new_flow_entrypoint`; I think that'll be a problem now that that storage is shared?

**Comment by tomastigera** [comment]
> Looks like `calico_tc_skb_accepted` accesses state->ct_result`_before_ jumping to`calico_tc_skb_new_flow_entrypoint`; I think that'll be a problem now that that storage is shared?

Yes, but since we had CT miss and hence it is a new flow, there is no data in that area which we might loose. Perhaps warrants a comment.

**Comment by marvin-tigera** [comment]
Removing "merge-when-ready" label due to new commits

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Review by fasaxc — APPROVED** [review]

**Pull request merged**
